DOCSP-37046: csot in v2 (#375)
mongodb/docs-golang: source/fundamentals/connections/connection-options.txt

       server documentation on
       :manual:`Write Concern options </reference/write-concern>`.

   * - **directConnection**
     - boolean
     - ``false``
     - Specifies whether to force dispatch **all** operations to the host
       specified in the connection URI.

For a full list of connection options, see the `ClientOptions API
documentation
<{+api+}/mongo/options#ClientOptions>`__.

.. _golang-timeout-setting:

Single Timeout Setting
----------------------

You can set a single ``Timeout`` option on your ``Client`` instance to
specify the amount of time that a single operation can take to execute.
Set a timeout by using the ``SetTimeout()`` method or specifying the
``timeoutMS`` option in your connection URI string. ``Database``, ``Collection``,
``Session``, ``ChangeStream``, and ``Bucket`` instances elsewhere in
your code inherit the ``Timeout`` option from ``Client`` if you do not
set a different timeout in a Context for specific operations.

If you pass a Context into an operation with a deadline, the driver uses
that Context deadline for the operation. If the context does not have a
deadline, the driver derives a new Context from the given Context by using

> the ``Timeout`` option set on the ``Client``.

Reviewer: This might be due to me not being very familiar with Go, but this paragraph is confusing to me. It seems a bit jarring to be talking about setting timeouts, but then the next paragraph is talking about using Context deadlines. Is a "deadline" the same thing as a timeout? If so I think we should stick with that wording. Happy to workshop some ways to rephrase this once I have a better understanding of what it's trying to say.

Author: Will clean this up (old text)

.. note:: Retries under Timeout Specification

   When using default settings, if you set a ``Timeout`` option on your ``Client``
   and your operation requires a retry, the driver retries the operation
   as many times as possible before the timeout expires. Once the timeout
   expires, the driver returns a timeout error. Versions 1.1 and later
   of the {+driver-short+} enable retryable reads and writes by default.
   See the Server manual for more information about :ref:`retryable
   reads <retryable-reads>` and :manual:`retryable writes </core/retryable-writes/>`.

Timeout Examples
~~~~~~~~~~~~~~~~

This section provides examples that demonstrate different ways to set the
``Timeout`` setting in your application.

Client Option
^^^^^^^^^^^^^

The following code shows how to set the ``Timeout`` option on a ``Client``
by using the ``SetTimeout()`` method:

.. code-block:: go

   opts := options.Client().SetTimeout(5 * time.Second)
   client, err := mongo.Connect(opts)

Connection String Option
^^^^^^^^^^^^^^^^^^^^^^^^

> The following example shows how you can set a single timeout with the
> URI option and execute an operation that inherits this setting:

Reviewer: ```suggestion
The following example shows how to set a single timeout with the
URI option. It then executes an operation that inherits this timeout.
```

.. code-block:: go

   uri := "mongodb://user:pass@sample.host:27017/?timeoutMS=5000"
   
   client, err := mongo.Connect(options.Client().ApplyURI(uri))
   
   ...

>    coll.InsertOne(context.Background(), doc)

Reviewer: S: Highlight line 1. And maybe line 7?

Multiple Timeouts
^^^^^^^^^^^^^^^^^

> The following example shows how you can set the ``Timeout`` option on a
> ``Client`` instance, then set a different timeout for an insert option
> within the Context passed to the ``InsertOne()`` method:

Reviewer: ```suggestion
The following example shows how to set the ``Timeout`` option on a
``Client`` instance. It then sets a different timeout
within the Context passed to the ``InsertOne()`` method:
```

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, timeout, customize connection

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

This guide explains several common MongoDB connection and authentication
options. You can pass the connection options as parameters of the connection
URI to specify the behavior of the client.

Options
-------

The following table describes the connection options that you can set in
your connection URI. Each entry provides the option name, value type,
default value, and a description of the option.

.. list-table::
   :header-rows: 1
   :widths: 34 10 12 44

   * - Option Name
     - Type
     - Default Value
     - Description

   * - **timeoutMS**
     - integer
     - ``null``
     - Specifies the number of milliseconds that a single operation run on the
       ``Client`` can take before returning a timeout error. Operations honor
       this setting only if there is no deadline on the operation Context.

   * - **connectTimeoutMS**
     - integer
     - ``30000``
     - Specifies the number of milliseconds to wait before timeout on a TCP

>        connection.

Author: Rewrote the desc to leave the acronym out (simplify)

       suitable server to execute an operation.

   * - **heartbeatFrequencyMS**
     - integer
     - ``10000``
     - Specifies the number of milliseconds to wait between periodic
       background server checks.

   * - **tls**
     - boolean
     - ``false``
     - Specifies whether to establish a Transport Layer Security (TLS)
       connection with the instance. This is automatically set to ``true``
       when using a DNS seedlist (SRV) in the connection string. You can
       override this behavior by setting the value to ``false``.

   * - **w**
     - string or integer
     - ``null``
     - Specifies the write concern. To learn more about values, see the
       server documentation on
       :manual:`Write Concern options </reference/write-concern>`.

   * - **directConnection**
     - boolean
     - ``false``
     - Specifies whether to force dispatch **all** operations to the host
       specified in the connection URI.

For a full list of connection options, see the `ClientOptions API
documentation
<{+api+}/mongo/options#ClientOptions>`__.

.. _golang-timeout-setting:

Single Timeout Setting
----------------------

> You can set a single ``Timeout`` option on your ``Client`` instance to

Author: The ``SetTimeout()`` method sets the ``Timeout`` struct field of the ``Client`` options struct.

> You can set a single ``Timeout`` option on your ``Client`` instance to

Reviewer: Ahh I see, that makes more sense. I think it's probably okay then, people more familiar with Go will likely understand that

       background server checks.

   * - **tls**
     - boolean
     - ``false``
     - Specifies whether to establish a Transport Layer Security (TLS)
       connection with the instance. This is automatically set to ``true``
       when using a DNS seedlist (SRV) in the connection string. You can
       override this behavior by setting the value to ``false``.

   * - **w**
     - string or integer
     - ``null``
     - Specifies the write concern. To learn more about values, see the
       server documentation on
       :manual:`Write Concern options </reference/write-concern>`.

   * - **directConnection**
     - boolean
     - ``false``
     - Specifies whether to force dispatch **all** operations to the host
       specified in the connection URI.

For a full list of connection options, see the `ClientOptions API
documentation
<{+api+}/mongo/options#ClientOptions>`__.

.. _golang-timeout-setting:

Single Timeout Setting
----------------------

You can set a single ``Timeout`` option on your ``Client`` instance to
specify the maximum amount of time that a single operation can take to
execute.

Set a timeout by calling the ``SetTimeout()`` method when specifying

> options for your ``Client`` instance orby specifying the

Reviewer: ```suggestion
options for your ``Client`` instance or by specifying the
```

``timeoutMS`` option in your connection URI. By default, all
``Database``, ``Collection``, ``Session``, ``ChangeStream``, and
``Bucket`` instances elsewhere in your application inherit the
``Timeout`` option from ``Client`` if you do not set a different timeout
on specific operations in the operation's Context.

If you set a timeout on a Context passed into an operation, the driver uses
that value for the operation. If you do not specify a Context timeout,
the operation Context derives the timeout from the ``Client`` instance.

.. note:: Retries under Timeout Specification

>    When using default settings, if you set a ``Timeout`` option on your ``Client``

Reviewer 2: This behavior is true for a client-level or operation-level (context.Context) CSOT. I would also clarify that this only happens if the server returns a "retryable" error, such as a socket exception.

https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.md#retryability

   and your operation requires a retry, the driver retries the operation
   as many times as possible before the timeout expires. Once the timeout
   expires, the driver returns a timeout error. Versions 1.1 and later
   of the {+driver-short+} enable retryable reads and writes by default.
   See the Server manual for more information about :ref:`retryable
   reads <retryable-reads>` and :manual:`retryable writes </core/retryable-writes/>`.

Timeout Examples
~~~~~~~~~~~~~~~~

This section provides examples that demonstrate different ways to set the
``Timeout`` setting in your application.

Client Option
^^^^^^^^^^^^^

The following code shows how to set the ``Timeout`` option on a ``Client``
by using the ``SetTimeout()`` method:

.. code-block:: go

   opts := options.Client().SetTimeout(5 * time.Second)
   client, err := mongo.Connect(opts)

> Connection String Option

Reviewer 2: Suggest specifying that the `timeoutMS` query parameter is a client-level timeout.

^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to set a single timeout by using the
``timeoutMS`` URI option. Then, the code executes an insert operation
that inherits the timeout:

.. code-block:: go
   :emphasize-lines: 1, 5

   uri := "mongodb://user:pass@sample.host:27017/?timeoutMS=5000"
   client, err := mongo.Connect(options.Client().ApplyURI(uri))
   
   ...
   coll.InsertOne(context.Background(), doc)

> Multiple Timeouts

Reviewer 2: Suggest renaming this section to "Operation Timeout"

^^^^^^^^^^^^^^^^^

The following example shows how to set the ``Timeout`` option on a
``Client`` instance. Then, the code sets a different timeout within the
Context passed to the ``InsertOne()`` method:

.. code-block:: go
   :emphasize-lines: 1, 5

   opts := options.Client().SetTimeout(5 * time.Second)

>    client, err := mongo.Connect(opts)

Reviewer 2: I don't think it's necessary to show this. For an operation-level timeout, it doesn't matter if a client-timeout is set or not, the operation timeout takes priority. Suggest removing this linea nd updating the description of this example:

```
The following example shows how to set an operation-level timeout, which 
takes priority over a client-level timeout if one is defined:
```
